frame/referenda/src/lib.rs

```diff
@@ -0,0 +1,1065 @@
+		let next_block = now + One::one();
+		let alarm_interval = T::AlarmInterval::get().max(One::one());
+		let when = (next_block + alarm_interval - One::one()) / alarm_interval * alarm_interval;
+
+		let maybe_result = T::Scheduler::schedule(
+			DispatchTime::At(when),
+			None,
+			128u8,
+			frame_system::RawOrigin::Root.into(),
+			MaybeHashed::Value(Call::one_fewer_deciding { track }.into()),
+		);
+		debug_assert!(
+			maybe_result.is_ok(),
+			"Unable to schedule a new alarm at #{:?} (now: #{:?})?!",
+			when,
+			now
+		);
+	}
+
+	/// Ensure that a `service_referendum` alarm happens for the referendum `index` at `alarm`.
+	///
+	/// This will do nothing if the alarm is already set.
+	///
+	/// Returns `false` if nothing changed.
+	fn ensure_alarm_at(
+		status: &mut ReferendumStatusOf<T>,
+		index: ReferendumIndex,
+		alarm: T::BlockNumber,
+	) -> bool {
+		if status.alarm.as_ref().map_or(true, |&(when, _)| when != alarm) {
+			// Either no alarm or one that was different
+			Self::ensure_no_alarm(status);
+			status.alarm = Self::set_alarm(Call::nudge_referendum { index }, alarm);
+			true
+		} else {
+			false
+		}
+	}
+
+	/// Advance the state of a referendum, which comes down to:
+	/// - If it's ready to be decided, start deciding;
+	/// - If it's not ready to be decided and non-deciding timeout has passed, fail;
+	/// - If it's ongoing and passing, ensure confirming; if at end of confirmation period, pass.
+	/// - If it's ongoing and not passing, stop confirning; if it has reached end time, fail.
+	///
+	/// Weight will be a bit different depending on what it does, but it's designed so as not to
+	/// differ dramatically, especially if `MaxQueue` is kept small. In particular _there are no
+	/// balance operations in here_.
+	///
+	/// In terms of storage, every call to it is expected to access:
+	/// - The scheduler, either to insert, remove or alter an entry;
+	/// - `TrackQueue`, which should be a `BoundedVec` with a low limit (8-16).
+	/// - `DecidingCount`.
+	///
+	/// Both of the two storage items will only have as many items as there are different tracks,
+	/// perhaps around 10 and should be whitelisted.
+	///
+	/// The heaviest branch is likely to be when a proposal is placed into, or moved within, the
+	/// `TrackQueue`. Basically this happens when a referendum is in the deciding queue and receives
+	/// a vote, or when it moves into the deciding queue.
+	fn service_referendum(
+		now: T::BlockNumber,
+		index: ReferendumIndex,
+		mut status: ReferendumStatusOf<T>,
+	) -> (ReferendumInfoOf<T>, bool, ServiceBranch) {
+		let mut dirty = false;
+		// Should it begin being decided?
+		let track = match Self::track(status.track) {
+			Some(x) => x,
+			None => return (ReferendumInfo::Ongoing(status), false, ServiceBranch::Fail),
+		};
+		let timeout = status.submitted + T::UndecidingTimeout::get();
+		// Default the alarm to the submission timeout.
+		let mut alarm = timeout;
+		let branch;
+		match &mut status.deciding {
```

Comment:
Maybe not a bad suggestion in this case, but note that patterns like this can be bad for `MaxEncodedLen` since we must assume the largest encoded size of the enum for all variants. So as long as the enum is not that variable, this is okay.